File: app/src/main/java/com/sjn/stamp/utils/SongStateHelper.java
package com.sjn.stamp.utils;


import android.app.Activity;
import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.drawable.AnimationDrawable;
import android.graphics.drawable.Drawable;
import android.support.v4.content.ContextCompat;
import android.support.v4.graphics.drawable.DrawableCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.PlaybackStateCompat;

import com.sjn.stamp.R;

public class SongStateHelper {

    private static final int STATE_INVALID = -1;
    private static final int STATE_NONE = 0;
    private static final int STATE_PLAYABLE = 1;
    private static final int STATE_PAUSED = 2;
    private static final int STATE_PLAYING = 3;

    private static ColorStateList sColorStatePlaying;
    private static ColorStateList sColorStateNotPlaying;

    public static Drawable getDrawableByState(Context context, int state) {
        if (sColorStateNotPlaying == null || sColorStatePlaying == null) {
            initializeColorStateLists(context);
        }

        switch (state) {
            case STATE_PLAYABLE:
                Drawable pauseDrawable = ContextCompat.getDrawable(context,
                        R.drawable.ic_play_arrow_black_36dp);
                DrawableCompat.setTintList(pauseDrawable, sColorStateNotPlaying);
                return pauseDrawable;
            case STATE_PLAYING:
                AnimationDrawable animation = (AnimationDrawable)
                        ContextCompat.getDrawable(context, R.drawable.ic_equalizer_white_36dp);
                DrawableCompat.setTintList(animation, sColorStatePlaying);
                animation.start();
                return animation;
            case STATE_PAUSED:
                Drawable playDrawable = ContextCompat.getDrawable(context,
                        R.drawable.ic_equalizer1_white_36dp);
                DrawableCompat.setTintList(playDrawable, sColorStatePlaying);
                return playDrawable;
            default:
                return null;
        }
    }

    public static int getMediaItemState(Activity activity, String mediaId, boolean isPlayable) {
        if (isPlayable) {
            return getMediaItemState(activity, mediaId);
        }
        return STATE_PLAYABLE;
    }

    private static int getMediaItemState(Activity activity, String mediaId) {
        int state = STATE_PLAYABLE;
        if (MediaIDHelper.isMediaItemPlaying(activity, mediaId)) {
            state = getStateFromController(activity);
        }
        return state;
    }

    private static void initializeColorStateLists(Context ctx) {
        sColorStateNotPlaying = ColorStateList.valueOf(ctx.getResources().getColor(
                R.color.media_item_icon_not_playing));
        sColorStatePlaying = ColorStateList.valueOf(ctx.getResources().getColor(
                R.color.media_item_icon_playing));
    }

    private static int getStateFromController(Activity activity) {
        MediaControllerCompat controller = MediaControllerCompat.getMediaController(activity);
        PlaybackStateCompat pbState = controller.getPlaybackState();
        if (pbState == null ||
                pbState.getState() == PlaybackStateCompat.STATE_ERROR) {
            return STATE_NONE;
        } else if (pbState.getState() == PlaybackStateCompat.STATE_PLAYING) {
            return STATE_PLAYING;
        } else {
            return STATE_PAUSED;
        }
    }
}

File: app/src/main/java/com/sjn/stamp/utils/SongStateHelper.kt
package com.sjn.stamp.utils


import android.app.Activity
import android.content.Context
import android.graphics.drawable.AnimationDrawable
import android.graphics.drawable.Drawable
import android.support.v4.content.ContextCompat
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.PlaybackStateCompat

import com.sjn.stamp.R

@Suppress("unused")
object SongStateHelper {

    private const val STATE_INVALID = -1
    private const val STATE_NONE = 0
    private const val STATE_PLAYABLE = 1
    private const val STATE_PAUSED = 2
    private const val STATE_PLAYING = 3

    fun getDrawableByState(context: Context, state: Int): Drawable? {

        when (state) {
            STATE_PLAYABLE -> {
                return ContextCompat.getDrawable(context,
                        R.drawable.ic_play_arrow_black_36dp)
            }
            STATE_PLAYING -> {
                val animation = ContextCompat.getDrawable(context, R.drawable.ic_equalizer_white_36dp) as AnimationDrawable?
                animation?.start()
                return animation
            }
            STATE_PAUSED -> {
                return ContextCompat.getDrawable(context,
                        R.drawable.ic_equalizer1_white_36dp)
            }
            else -> return null
        }
    }

    fun getMediaItemState(activity: Activity, mediaId: String, isPlayable: Boolean): Int =
            if (isPlayable) getMediaItemState(activity, mediaId) else STATE_PLAYABLE

    private fun getMediaItemState(activity: Activity, mediaId: String): Int =
            if (MediaIDHelper.isMediaItemPlaying(activity, mediaId)) getStateFromController(activity) else STATE_PLAYABLE

    private fun getStateFromController(activity: Activity): Int {
        val pbState = MediaControllerCompat.getMediaController(activity).playbackState
        return if (pbState == null || pbState.state == PlaybackStateCompat.STATE_ERROR) {
            STATE_NONE
        } else if (pbState.state == PlaybackStateCompat.STATE_PLAYING) {
            STATE_PLAYING
        } else {
            STATE_PAUSED
        }
    }
}
